src/treehash.rs

```diff
@@ -0,0 +1,537 @@
+    // u16::MAX - 1 if the pair has been seen once
+    // u16::MAX - 2 if the pair has been seen at least twice (this makes it a
+    // candidate for memoization)
+    pairs: Vec<u16>,
+}
+
+const NOT_VISITED: u16 = u16::MAX;
+const SEEN_ONCE: u16 = u16::MAX - 1;
+const SEEN_MULTIPLE: u16 = u16::MAX - 2;
+
+impl TreeCache {
+    /// Get cached hash and its associated cost (if present).
+    pub fn get(&self, n: NodePtr) -> Option<(&[u8; 32], Cost)> {
+        // We only cache pairs (for now)
+        if !matches!(n.object_type(), ObjectType::Pair) {
+            return None;
+        }
+
+        let idx = n.index() as usize;
+        let slot = *self.pairs.get(idx)?;
+        if slot >= SEEN_MULTIPLE {
+            return None;
+        }
+        Some((&self.hashes[slot as usize], self.costs[slot as usize]))
+    }
+
+    /// Insert a cached hash with its associated cost. If the cache is full we
+    /// ignore the insertion.
+    pub fn insert(&mut self, n: NodePtr, hash: &[u8; 32], cost: Cost) {
+        // If we've reached the max size, just ignore new cache items
+        if self.hashes.len() == SEEN_MULTIPLE as usize {
+            return;
+        }
+
+        if !matches!(n.object_type(), ObjectType::Pair) {
+            return;
+        }
+
+        let idx = n.index() as usize;
+        if idx >= self.pairs.len() {
+            self.pairs.resize(idx + 1, NOT_VISITED);
+        }
+
+        let slot = self.hashes.len();
+        self.hashes.push(*hash);
+        self.costs.push(cost);
+        self.pairs[idx] = slot as u16;
+    }
+
+    /// mark the node as being visited. Returns true if we need to
+    /// traverse visitation down this node.
+    fn visit(&mut self, n: NodePtr) -> bool {
+        if !matches!(n.object_type(), ObjectType::Pair) {
+            return false;
+        }
+        let idx = n.index() as usize;
+        if idx >= self.pairs.len() {
+            self.pairs.resize(idx + 1, NOT_VISITED);
+        }
+        if self.pairs[idx] > SEEN_MULTIPLE {
+            self.pairs[idx] -= 1;
+        }
+        self.pairs[idx] == SEEN_ONCE
+    }
+
+    pub fn should_memoize(&mut self, n: NodePtr) -> bool {
+        if !matches!(n.object_type(), ObjectType::Pair) {
+            return false;
+        }
+        let idx = n.index() as usize;
+        if idx >= self.pairs.len() {
+            false
+        } else {
+            self.pairs[idx] <= SEEN_MULTIPLE
+        }
+    }
+
+    pub fn visit_tree(&mut self, a: &Allocator, node: NodePtr) {
+        if !self.visit(node) {
+            return;
+        }
+        let mut nodes = vec![node];
+        while let Some(n) = nodes.pop() {
+            let SExp::Pair(left, right) = a.sexp(n) else {
+                continue;
+            };
+            if self.visit(left) {
+                nodes.push(left);
+            }
+            if self.visit(right) {
+                nodes.push(right);
+            }
+        }
+    }
+}
```

Comment:
I think all code related to the cached version should be removed until we have a good solution. I suspect it will need to look very different from this